crates/chain/src/tx_graph.rs

```diff
@@ -232,176 +482,207 @@ impl TxGraph {
-            outpoint.txid,
-            TxNode::Partial([(outpoint.vout, txout)].into()),
-        );
-        self.determine_additions(&update)
+        additions
     }
 }
 
-impl TxGraph {
-    /// The transactions spending from this output.
-    ///
-    /// `TxGraph` allows conflicting transactions within the graph. Obviously the transactions in
-    /// the returned set will never be in the same active-chain.
-    pub fn outspends(&self, outpoint: OutPoint) -> &HashSet<Txid> {
-        self.spends.get(&outpoint).unwrap_or(&self.empty_outspends)
+impl<A: BlockAnchor> TxGraph<A> {
+    /// Get all heights that are relevant to the graph.
+    pub fn relevant_heights(&self) -> BTreeSet<u32> {
+        self.anchors
+            .iter()
+            .map(|(a, _)| a.anchor_block().height)
+            .collect()
     }
 
-    /// Iterates over the transactions spending from `txid`.
+    /// Determines whether a transaction of `txid` is in the best chain.
     ///
-    /// The iterator item is a union of `(vout, txid-set)` where:
-    ///
-    /// - `vout` is the provided `txid`'s outpoint that is being spent
-    /// - `txid-set` is the set of txids spending the `vout`.
-    pub fn tx_outspends(
+    /// TODO: Also return conflicting tx list, ordered by last_seen.
+    pub fn try_get_chain_position<C>(
         &self,
+        chain: C,
         txid: Txid,
-    ) -> impl DoubleEndedIterator<Item = (u32, &HashSet<Txid>)> + '_ {
-        let start = OutPoint { txid, vout: 0 };
-        let end = OutPoint {
-            txid,
-            vout: u32::MAX,
+    ) -> Result<Option<ObservedIn<&A>>, C::Error>
+    where
+        C: ChainOracle,
+    {
+        let (tx_node, anchors, &last_seen) = match self.txs.get(&txid) {
+            Some((tx, anchors, last_seen)) if !(anchors.is_empty() && *last_seen == 0) => {
+                (tx, anchors, last_seen)
+            }
+            _ => return Ok(None),
         };
-        self.spends
-            .range(start..=end)
-            .map(|(outpoint, spends)| (outpoint.vout, spends))
-    }
 
-    /// Iterate over all partial transactions (outputs only) in the graph.
-    pub fn partial_transactions(&self) -> impl Iterator<Item = (Txid, &BTreeMap<u32, TxOut>)> {
-        self.txs.iter().filter_map(|(txid, tx)| match tx {
-            TxNode::Whole(_) => None,
-            TxNode::Partial(partial) => Some((*txid, partial)),
-        })
+        for anchor in anchors {
+            if chain.is_block_in_best_chain(anchor.anchor_block())? {
+                return Ok(Some(ObservedIn::Block(anchor)));
+            }
+        }
+
+        // The tx is not anchored to a block which is in the best chain, let's check whether we can
+        // ignore it by checking conflicts!
+        let tx = match tx_node {
+            TxNodeInternal::Whole(tx) => tx,
+            TxNodeInternal::Partial(_) => {
+                // [TODO] Unfortunately, we can't iterate over conflicts of partial txs right now!
```

Comment:
This doesn't seem unfortunate? How would this be possible even?
Remove these `TODO`.

Reply:
Are you saying partial transactions cannot have conflicts?

Reply:
How can you possible infer the conflicts of a partial transaction since you don't know any of the parent nodes?

Reply:
Lol, that is a good point.